File: Week3/Code/get_TreeHeight.py
#!usr/bin/env python3
"""Python version of get_TreeHeight.R. calculate the tree heights from an input csv file and save the results in Results directory.  """
__appname__ = "Get TreeHeight"
__author__ = "Xiang Li(xiang.li419@imperial.ac.uk)"
__version__ = "0.0.1"
__license__ = "none"

###imports
import sys
import numpy as np 
import csv
import pandas as pd
import re #regular expressions
###globals


###functions
def TreeHeight(m):#m[0]: species;    m[1]: distance in meter    m[2]:degrees
    """calculates tree heights.
    input: m where m[:,0]: species;    m[:,1]: distance in meter    m[:,2]:degrees
    output: a list contians all heights """
    radians = m[:,2].astype("float") * np.pi /180 ##coerce from character to numeric type for calculation
    heights =m[:,1].astype("float") * np.tan(radians) 
    return heights


def main(argv):
    """main entry of the program"""
    trees = pd.read_csv(argv[1])#read file
    trees_m = trees.values#convert to array
    heights = TreeHeight(trees_m)#calculate heights
    trees["Tree.Height.m"] = heights
    filename = "../Results/" + re.split("\\.|/", argv[1])[4] + "_treeheights.csv"
    print ("writng to result file...")
    trees.to_csv(filename)
    return 0

if __name__ == "__main__":
    status = main(sys.argv)
    sys.exit(status)

File: Week3/Code/get_TreeHeight.R
# This function calculates heights of trees given distance of each tree 
# from its base and angle to its top, using  the trigonometric formula 
#
# height = distance * tan(radians)
#
# ARGUMENTS
# a .csv file which includes follows:
# degrees:   The angle of elevation of tree
# distance:  The distance from base of tree (e.g., meters)
#
# OUTPUT
# The heights of the tree, same units as "distance"
##############
#take a csv file name from the command line(e.g trees.csv)
#Loads the .csv file and calculates tree heights for all trees in the data. 
#Creates a csv output file named as InputFileName_treeheights.csv in results that contains the calculated 
#tree heights along with the original data 

rm(list = ls()) #clean current obejcts in workspace

###functions
TreeHeight <- function(m){ #m[,1]:species    m[,2]:distances in meter    m[,3]: degrees
#apply calculations on each column
  radians <- as.numeric(m[,3]) * pi /180 ##coerce from character to numeric type for calculation
  height <-as.numeric(m[,2]) * tan(radians)
  return (height) #return a vector of heights
}

##handling command line argument
args <- commandArgs(trailingOnly = T) 
if(length(args) == 0){
  stop("No input file.")
}

trees.df <- read.csv(args[1]) # read csv file
trees.m <- as.matrix(trees.df)#matrix using matrix instead of data.frame because it is faster when lager number of mathematical 
#calculations

trees.df$Tree.Height.m <-TreeHeight(trees.m)

#split argument using / and . to get input filename; using paste() to get output filename
filename <- paste("../Results/", strsplit(args[1], "/|\\.")[[1]][5], "_treeheights.csv", sep = "")

write.csv(file = filename, trees.df) #write results
print(paste("results are written into", filename))



